volume/volumefs/filesystem.go

```diff
@@ -0,0 +1,93 @@
+// a snapshot optimized block size defined by the filesystem itself. This "snapshot
+// block" size will typically be a multiple of the volume block size.  The filesystem
+// represents snapshot blocks as "files" named for their snapshot block address. Files
+// are organized in a directory hierarchy to address scale. See layout.go for details.
+//
+// The first snapshot of the volume backs up all volume blocks (typically only the ones
+// written out at this time). On subsequent snapshots the filesystem supports volumes
+// with underlying support for change-block-tracking: only the snapshot blocks changed
+// since the previous volume snapshot need be considered, as the filesystem recovers
+// its previous state from the repository.
+package volumefs
+
+import (
+	"fmt"
+	"path"
+	"time"
+
+	"github.com/kopia/kopia/fs"
+	"github.com/kopia/kopia/repo"
+	"github.com/kopia/kopia/repo/logging"
+	"github.com/kopia/kopia/snapshot"
+	"github.com/kopia/kopia/volume"
+)
+
+var log = logging.GetContextLoggerFunc("volume/filesystem")
+
+// package errors.
+var (
+	ErrOutOfRange  = fmt.Errorf("block address out of range")
+	ErrInvalidArgs = fmt.Errorf("invalid arguments")
+)
+
+// FilesystemArgs contains arguments to create a filesystem
+type FilesystemArgs struct {
+	// The repository.
+	Repo repo.Repository
+	// The volume manager.
+	VolumeManager volume.Manager
+	// The identifier of the volume being backed up or restored.
+	VolumeID string
+	// The identifier of the volume snapshot being backed up or restored.
+	VolumeSnapshotID string
+	// Profile containing location and credential information needed to access the volume.
+	VolumeAccessProfile interface{}
+}
+
+// Validate checks for required fields
+func (a *FilesystemArgs) Validate() error {
+	if a.Repo == nil || a.VolumeManager == nil || a.VolumeID == "" || a.VolumeSnapshotID == "" || a.VolumeAccessProfile == nil {
+		return ErrInvalidArgs
+	}
+
+	return nil
+}
+
+// SourceInfo generates a snapshotSourceInfo
+func (a *FilesystemArgs) SourceInfo() snapshot.SourceInfo {
+	return snapshot.SourceInfo{
+		Path:     path.Join("/volumefs", a.VolumeManager.Type(), a.VolumeID),
+		Host:     a.Repo.Hostname(),
+		UserName: a.Repo.Username(),
+	}
+}
+
+// Filesystem is a pseudo-filesystem mapping a provider volume block
+// address space to the snapshot block address space.
+type Filesystem struct {
+	FilesystemArgs
+	layoutProperties
+	blockReader       volume.BlockReader
+	epoch             time.Time // all changes stamped with this time
+	rootDir           *dirMeta
+	previousRootEntry fs.Directory
+	logger            logging.Logger
+}
+
+// New returns a new volume filesystem
+func New(args *FilesystemArgs) (*Filesystem, error) {
+	if err := args.Validate(); err != nil {
+		return nil, err
+	}
+
+	f := &Filesystem{}
+	f.FilesystemArgs = *args
+	f.setDefaultLayoutProperties() // block size may be reset from previous repo snapshot
```

Comment:
After discussion with @julio-lopez we should parameterize the layout properties for experimentation